collector/exchange.go

```diff
@@ -0,0 +1,618 @@
+			labelName,
+		)
+		ch <- prometheus.MustNewConstMetric(
+			c.InternalLargestDeliveryQueueLength,
+			prometheus.GaugeValue,
+			queue.InternalLargestDeliveryQueueLength,
+			labelName,
+		)
+		ch <- prometheus.MustNewConstMetric(
+			c.PoisonQueueLength,
+			prometheus.GaugeValue,
+			queue.PoisonQueueLength,
+			labelName,
+		)
+	}
+	return nil
+}
+
+// Perflib: [19430] MSExchange WorkloadManagement Workloads
+type perflibWorkloadManagementWorkloads struct {
+	Name string
+
+	ActiveTasks    float64 `perflib:"ActiveTasks"`
+	CompletedTasks float64 `perflib:"CompletedTasks"`
+	QueuedTasks    float64 `perflib:"QueuedTasks"`
+	YieldedTasks   float64 `perflib:"YieldedTasks"`
+	IsActive       float64 `perflib:"Active"`
+}
+
+func (c *exchangeCollector) collectWorkloadManagementWorkloads(ctx *ScrapeContext, ch chan<- prometheus.Metric) error {
+	var data []perflibWorkloadManagementWorkloads
+	if err := unmarshalObject(ctx.perfObjects["MSExchange WorkloadManagement Workloads"], &data); err != nil {
+		return err
+	}
+
+	for _, instance := range data {
+		labelName := toLabelName(instance.Name)
+		if strings.HasSuffix(labelName, "_total") {
+			continue
+		}
+		ch <- prometheus.MustNewConstMetric(
+			c.ActiveTasks,
+			prometheus.GaugeValue,
+			instance.ActiveTasks,
+			labelName,
+		)
+		ch <- prometheus.MustNewConstMetric(
+			c.CompletedTasks,
+			prometheus.CounterValue,
+			instance.CompletedTasks,
+			labelName,
+		)
+		ch <- prometheus.MustNewConstMetric(
+			c.QueuedTasks,
+			prometheus.CounterValue,
+			instance.QueuedTasks,
+			labelName,
+		)
+		ch <- prometheus.MustNewConstMetric(
+			c.YieldedTasks,
+			prometheus.CounterValue,
+			instance.YieldedTasks,
+			labelName,
+		)
+		ch <- prometheus.MustNewConstMetric(
+			c.IsActive,
+			prometheus.GaugeValue,
+			instance.IsActive,
+			labelName,
+		)
+	}
+
+	return nil
+}
+
+// [29240] MSExchangeAutodiscover
+type perflibAutodiscover struct {
+	RequestsPerSec float64 `perflib:"Requests/sec"`
+}
+
+func (c *exchangeCollector) collectAutoDiscover(ctx *ScrapeContext, ch chan<- prometheus.Metric) error {
+	var data []perflibAutodiscover
+	if err := unmarshalObject(ctx.perfObjects["MSExchangeAutodiscover"], &data); err != nil {
+		return err
+	}
+	for _, autodisc := range data {
+		ch <- prometheus.MustNewConstMetric(
+			c.AutodiscoverRequestsPerSec,
+			prometheus.CounterValue,
+			autodisc.RequestsPerSec,
+		)
+	}
+	return nil
+}
+
+// desc creates a new prometheus description
+func desc(metricName string, description string, labels ...string) *prometheus.Desc {
+	return prometheus.NewDesc(
+		prometheus.BuildFQName(Namespace, "exchange", metricName),
+		description,
+		labels,
+		nil,
+	)
+}
+
+// toLabelName converts strings to lowercase and replaces all whitespace and dots with underscores
+func toLabelName(name string) string {
+	s := strings.ReplaceAll(strings.Join(strings.Fields(strings.ToLower(name)), "_"), ".", "_")
+	s = strings.ReplaceAll(s, "__", "_")
+	return s
+}
```

Comment:
Since these have pretty generic names and the `collector` package namespace is shared with all other collectors, could they either be renamed to something more specific, or made methods on the collector object? (Eg `func (c *exchangeCollector) desc(...)` rather than free functions)

Reply:
I'll fix the functions. The toLabelName is probably not the first function that does something like this and would probably be better implemented in a general way with tests and such.

The metric names/types are confusing, particularly when Microsoft has been consistently inconsistent when naming things. I'll go over them again

Can I trust the perflib type reporting? Is something called a _per_sec_ but a _counter_-type always a _total_

Reply:
Here is a better view of the metrics if and their types. 
Let me know if you see any errors.

https://gist.github.com/eikaas/8bb26d397b70e049c4339bd299e6c568

Reply:
Yeah, it is frustrating, I know... I don't think we've run into any cases where the types lie, though. So if perflib reports it as a counter, it should be correct, even/especially if the name is something like per second.
For the vast majority of cases, things called per sec has been a counter, and things called something like average latency or such has been a counter of number of occurences/events total, and a base counter with the total time. This isn't amazingly documented in our perflib wrapper, but when there's a base/value pair in a counter you can get the base by throwing a suffix `_Base` on it, so eg `perflib:"RPC Averaged Latency"` likely also has a `perflib:"RPC Averaged Latency_Base"`, where the first is the total time, and the second the total number of rpc:s.